### PR TITLE
SAMZA-2047: Fixing udfs that return samzasqlrelrecord type and other such types to work with joins.

### PR DESCRIPTION
**Pull request by atoomula**

### CODE REVIEW

File: samza-sql/src/main/java/org/apache/samza/sql/fn/BuildOutputRecordUdf.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ *     field31:{
+ *       field311,
+ *       field312
+ *     },
+ *     field32
+ *   }
+ * };
+ * It could be built in the select statement as:
+ * select obj1 as field1,
+ *        BuildSamzaSqlRelRecord("field21", obj21, "field22", obj22) as field2,
```
Comment (vjagadish1989) [inline samza-sql/src/main/java/org/apache/samza/sql/fn/BuildOutputRecordUdf.java]: Should the java-doc reflect the UDF name? BuildSamzaSqlRelRecord vs BuildOutputRecordUdf

File: samza-sql/src/test/java/org/apache/samza/sql/fn/TestBuildOutputRecordUdf.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    SamzaSqlRelRecord expectedRecord =
+        new SamzaSqlRelRecord(Arrays.asList("k1", "k2"),
+            Arrays.asList("v1", nestedSamzaSqlRelRecord));
+
+    Assert.assertEquals(actualRecord.getFieldNames(), expectedRecord.getFieldNames());
+    Assert.assertEquals(actualRecord.getFieldValues(), expectedRecord.getFieldValues());
+  }
+
+  @Test(expected = NullPointerException.class)
+  public void testNullArgs() {
```
Comment (vjagadish1989) [inline samza-sql/src/test/java/org/apache/samza/sql/fn/TestBuildOutputRecordUdf.java]: nice set of tests!

File: samza-sql/src/test/java/org/apache/samza/sql/testutil/SamzaSqlTestConfig.java [hunk 3]
```diff
@@ -95,7 +96,7 @@
         ConfigBasedUdfResolver.class.getName());
     staticConfigs.put(configUdfResolverDomain + ConfigBasedUdfResolver.CFG_UDF_CLASSES, Joiner.on(",")
         .join(MyTestUdf.class.getName(), RegexMatchUdf.class.getName(), FlattenUdf.class.getName(),
-            MyTestArrayUdf.class.getName()));
+            MyTestArrayUdf.class.getName(), BuildOutputRecordUdf.class.getName()));
```
Comment (weiqingy) [inline samza-sql/src/test/java/org/apache/samza/sql/testutil/SamzaSqlTestConfig.java]: In Shell, when user performs "show functions", Shell will show all the details of UDFs supported by SamzaSQL. Currently we manually register the UDF info in [listFunctions](https://github.com/apache/samza/blob/master/samza-sql-shell/src/main/java/org/apache/samza/sql/client/impl/SamzaExecutor.java#L323). I added TODO to fix this and make it automatically achieve all UDF details, for now, shall we still manually register the UDF info in  [listFunctions](https://github.com/apache/samza/blob/master/samza-sql-shell/src/main/java/org/apache/samza/sql/client/impl/SamzaExecutor.java#L323)?

Comment (atoomula) [reply samza-sql/src/test/java/org/apache/samza/sql/testutil/SamzaSqlTestConfig.java]: These are added explicitly for our integration tests. For shell, in future, we will have to add a way to detect all the udfs.

File: samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java [hunk 4]
```diff
@@ -280,7 +280,7 @@ private void validateJoinKeys(RexInputRef ref) {
     if (sqlTypeName != SqlTypeName.BOOLEAN && sqlTypeName != SqlTypeName.TINYINT && sqlTypeName != SqlTypeName.SMALLINT
         && sqlTypeName != SqlTypeName.INTEGER && sqlTypeName != SqlTypeName.CHAR && sqlTypeName != SqlTypeName.BIGINT
         && sqlTypeName != SqlTypeName.VARCHAR && sqlTypeName != SqlTypeName.DOUBLE && sqlTypeName != SqlTypeName.FLOAT
-        && sqlTypeName != SqlTypeName.ANY) {
+        && sqlTypeName != SqlTypeName.ANY && sqlTypeName != SqlTypeName.OTHER) {
```
[the remark this answers is not in the document]
Comment (atoomula) [reply samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java]: Good point. I will leave these validations up to the table to validate this as it is the one which implements the gets and puts.

**Pull request closed without merging**